src/wpanctl/tool-cmd-commissioner.c

```diff
@@ -0,0 +1,367 @@
+
+            ret = lookup_dbus_name_from_interface(interface_dbus_name, gInterfaceName);
+            if (ret != 0) {
+                print_error_diagnosis(ret);
+                goto bail;
+            }
+            snprintf(path,
+                     sizeof(path),
+                     "%s/%s",
+                     WPANTUND_DBUS_PATH,
+                     gInterfaceName);
+
+            message = dbus_message_new_method_call(
+                interface_dbus_name,
+                path,
+                WPANTUND_DBUS_APIv1_INTERFACE,
+                WPANTUND_IF_CMD_JOINER_ADD
+                );
+
+            if (psk) {
+                uint8_t psk_bytes[psk_len];
+                memset(psk_bytes, '\0', psk_len+1);
+
+                memcpy(psk_bytes, psk, psk_len);
+                char *psk = psk_bytes;
+
+                dbus_message_append_args(
+                    message,
+                    DBUS_TYPE_STRING, &psk,
+                    DBUS_TYPE_INVALID
+                    );
+
+                dbus_message_append_args(
+                    message,
+                    DBUS_TYPE_UINT32, &joiner_timeout,
+                    DBUS_TYPE_INVALID
+                    );
+
+                if (ext_addr) {
+                    uint8_t addr_bytes[EXT_ADDRESS_LENGTH];
+                    memset(addr_bytes, 0, EXT_ADDRESS_LENGTH);
+                    uint8_t *p_addr_bytes = addr_bytes;
+                    int length = parse_string_into_data(addr_bytes,
+                                                        EXT_ADDRESS_LENGTH,
+                                                        ext_addr);
+                    assert(length == EXT_ADDRESS_LENGTH);
+                    dbus_message_append_args(
+                        message,
+                        DBUS_TYPE_ARRAY, DBUS_TYPE_BYTE, &p_addr_bytes, length,
+                        DBUS_TYPE_INVALID
+                        );
+                }
+
+                reply = dbus_connection_send_with_reply_and_block(
+                    connection,
+                    message,
+                    timeout,
+                    &error
+                    );
+
+                if (!reply) {
+                    fprintf(stderr, "%s: error: %s\n", argv[0], error.message);
+                    ret = ERRORCODE_TIMEOUT;
+                    goto bail;
+                }
+
+                dbus_message_get_args(reply, &error,
+                                      DBUS_TYPE_INT32, &ret,
+                                      DBUS_TYPE_INVALID
+                                      );
+                if (!ret) {
+                    fprintf(stderr, "Joiner added.\n");
+                } else {
+                    fprintf(stderr, "%s failed with error %d. %s\n", argv[0], ret, wpantund_status_to_cstr(ret));
+                    print_error_diagnosis(ret);
+                }
+            }
+        }
+    }
+
+
+bail:
+    if (connection)
+        dbus_connection_unref(connection);
```

Comment:
Due to the potent example of the ["goto fail"](https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/) bug, we now require curly braces for `if` (and all similar constructs) in all new code.